parlai/agents/transformer/modules.py

```diff
@@ -999,12 +1063,48 @@ def prepare_head(tensor):
             # key and value are the same, but query differs
             # self attention
             value = key
-        _, key_len, dim = key.size()
+        _, _, dim = key.size()
 
         q = prepare_head(self.q_lin(query))
         k = prepare_head(self.k_lin(key))
         v = prepare_head(self.v_lin(value))
 
+        # Prepend incremental states
```

Comment:
This is large block of new code. Can you describe it a bit more

Reply:
Yes - described